### ISSUE TITLE
GDAL provider ignores proxy settings

### ISSUE THREAD

**Issue by kannes** [qgis/QGIS#51956]
### What is the bug or the crash?

When a proxy is set in _Options_ -> _Network_, QGIS and its tools should use this proxy to ensure connectivity, prevent data leakage, etc. The GDAL provider is ignoring these settings though, it's calls to native GDAL/OGR tools happen without the configured proxy settings.

I realize that this is probably pretty complicated and hairy (one would need to explicitly pass proxy options to GDAL), but it caught me by surprise when things did not work in a workflow where at some stage `ogr2ogr` was called on a WFS source which was inaccessible because of this. I think that users would expect proxy settings to be honored by core providers.

### Steps to reproduce the issue

1. Add a WFS layer to your QGIS project, e.g. `ms:inv80_lgeomorph` from `http://maps.zh.ch/wfs/FnsNSWFS` (random result from web search)
1. Go to _Options_ -> _Network_
2. Enable _Use Proxy for Web Access_
3. Choose _HttpProxy_ or something else where you can specify a proxy
4. Set the host to `localhost` and the port to a port where you do **not** have a proxy running (e.g. `34567` probably)
5. (_Don't_ test this right now, Processing would crash when trying to open the tool dialog in the next step (#TODO): If you would reload the project now, QGIS should not be able to load the layer, this shows that the proxy setting is correctly applied and the proxy is not working, as intended.)
6. Execute the _GDAL_ -> _Vector information_ tool (`gdal:ogrinfo`) for the layer
7. The tool will run and finish without any issues, providing the remote information of the layer as it bypasses the proxy settings and fetches them from the WFS

The tool should not bypass the proxy settings. For the test case, it should not be able to access a remote source.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS-Version | 3.28.3-Firenze | QGIS-Codeversion | c12bcb2f76
-- | -- | -- | --
Qt-Version | 5.15.3
Python-Version | 3.10.6
GDAL-Version | 3.4.1
PROJ-Version | 8.2.1
EPSG-Registraturdatenbankversion | v10.041 (2021-12-03)
GEOS-Version | 3.10.2-CAPI-1.16.0
SQLite-Version | 3.37.2
PDAL-Version | 2.3.0
PostgreSQL-Client-Version | unknown
SpatiaLite-Version | 5.0.1
QWT-Version | 6.1.4
QScintilla2-Version | 2.11.6
BS-Version | Ubuntu 22.04.1 LTS
  |   |   |  
Aktive Python-Erweiterungen
Qgis2threejs | 2.7.1
QuickOSM | 2.1.1
processing | 2.12.99
sagaprovider | 2.12.99
grassprovider | 2.12.99
MetaSearch | 0.3.6
db_manager | 0.1.20

</body></html>QGIS-Version
3.28.3-Firenze
QGIS-Codeversion
[c12bcb2f76](https://github.com/qgis/QGIS/commit/c12bcb2f76)
Qt-Version
5.15.3
Python-Version
3.10.6
GDAL-Version
3.4.1
PROJ-Version
8.2.1
EPSG-Registraturdatenbankversion
v10.041 (2021-12-03)
GEOS-Version
3.10.2-CAPI-1.16.0
SQLite-Version
3.37.2
PDAL-Version
2.3.0
PostgreSQL-Client-Version
unknown
SpatiaLite-Version
5.0.1
QWT-Version
6.1.4
QScintilla2-Version
2.11.6
BS-Version
Ubuntu 22.04.1 LTS

Aktive Python-Erweiterungen
Qgis2threejs
2.7.1
QuickOSM
2.1.1
processing
2.12.99
sagaprovider
2.12.99
grassprovider
2.12.99
MetaSearch
0.3.6
db_manager
0.1.20

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

The same happens with the GRASS provider, see #TODO

**Comment by jef-n** [comment]
duplicate of #14949

**Comment by kannes** [comment]
Hm, are you sure? That one seems to have been about layer sources in general, I had no issues there and my test setup did successfully fail to access WFS layers if a broken proxy was set.
This one is about calls to the executable binaries, e.g. `ogr2ogr`, by the processing tools.

**Comment by pathmapper** [comment]
Just hit this issue, it's not possible to add a COG layer (which uses GDAL) behind proxy:

Test case:
https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/36/Q/WD/2020/7/S2A_36QWD_20200701_0_L2A/TCI.tif

![grafik](https://github.com/qgis/QGIS/assets/20856381/938074d2-52bc-4141-88d4-8eb70872b3a4)

```
2024-02-06T11:07:12     INFO    Remote layer : loading /vsicurl/https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/36/Q/WD/2020/7/S2A_36QWD_20200701_0_L2A/TCI.tif, please wait …
2024-02-06T11:13:03     CRITICAL    Unsupported Data Source : /vsicurl/https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/36/Q/WD/2020/7/S2A_36QWD_20200701_0_L2A/TCI.tif is not a supported raster data source
             HTTP response code on https://sentinel-cogs.s3.us-west-2.amazonaws.com/sentinel-s2-l2a-cogs/36/Q/WD/2020/7/S2A_36QWD_20200701_0_L2A/TCI.tif: 0
```

---

This was reported before in https://github.com/qgis/QGIS/issues/42870 (which is closed but not fixed).

I think also related to https://github.com/qgis/QGIS/issues/53351, from @rouault's comment in https://github.com/qgis/QGIS/issues/53351#issuecomment-1593655413:

> Those requests are issued by GDAL itself, using curl.
> To be able to log such requests, it would require changes both in GDAL (similar to the CPLHTTPPushFetchCallback() infrastrcture that was added for the vector side) and QGIS. Not a trivial change

---

@kannes for the COG issue it's possible to work around with setting the env variable `https_proxy`. Likely it solves also your reported issue with processing - but I do agree that this should work out-of-the-box.

**Comment by m-kuhn** [comment]
I see two different approaches:

 - GDAL callbacks as discussed in #53351, with the advantage that we are able to log requests and have a fine control over the proxy (e.g. use different proxies for local network vs. internet connections).
 - Setting GDAL environment variables (`GDAL_HTTP_PROXY` etc), with the advantage, that this would also work for subprocesses (e.g. processing) and wouldn't require any changes to GDAL.

Especially (but not only) for the second scenario: we should also allow forwarding the "system proxy", which would require probing for the proxy using [`systemProxyForQuery`](https://doc.qt.io/qt-6/qnetworkproxyfactory.html#systemProxyForQuery). Another critical point is username/password, which should ideally not be available through environment variables, so this should be left aside and let the user at least use `GDAL_PROXY_AUTH` for ntlm etc as secure alternatives.

Opinions on this (CC @rouault )

**Comment by rouault** [comment]
> Opinions on this (CC [@rouault](https://github.com/rouault) )

I had thought about that issue some time ago and was undecisive of the best approach. The callback approach could go very complicated because it would basically meaning replicating all the API surface of libcurl that GDAL uses, and this is not trivial, and managing to replicate it in terms of Qt Networking. Once that is done, proxying is for free. Apart for external processes indeed

The second alternative is probably more workable, if we can always mirror system parameters in terms of curl proxy parameters.

**Comment by pathmapper** [comment]
In general, I believe that every network request executed by QGIS should be logged in the ‘Debugging/Development Tools’ to ensure transparency and facilitate debugging (this applies not only to GDAL processing provider, but also to plugins, for example).

That would be option 1.

If this is not technically possible or desirable:

Am I correct in understanding that with option 2, the user only configures their proxy settings under 

`Settings -> Options-> Network -> Use proxy for web access `

and that the environment variables are set automatically in the background, i.e. **NOT** via

`Settings -> Options -> System -> Environment `

?

**Comment by m-kuhn** [comment]
@pathmapper option 1 would definitely be cleaner in theory but a lot more complex in practice (and almost impossible for subprocesses)
For option 2 you are correct, the environment variables will be set automatically in the background

**Comment by pathmapper** [comment]
@m-kuhn as discussed, option 2 might also solve the problem for MetaSearch by setting the relevant environment variables: https://github.com/qgis/QGIS/issues/53159
